File: backend/security/security_engine.py
import time
from collections import defaultdict, deque

_abuse: dict[str, int] = defaultdict(int)
_recent: dict[str, deque[float]] = defaultdict(deque)


def validate_secure_action(action: str, actor_id: str, session_id: str, device_id: str, timestamp_ms: int, idempotency_key: str = "") -> tuple[bool, str]:
    if not action or not actor_id or not session_id or not device_id:
        return False, "missing_context"
    now_ms = int(time.time() * 1000)
    if abs(now_ms - int(timestamp_ms or 0)) > 5 * 60 * 1000:
        return False, "stale_action"
    key = idempotency_key or f"{action}:{actor_id}:{timestamp_ms//1000}"
    q = _recent[key]
    while q and time.time() - q[0] > 60:
        q.popleft()
    if q:
        _abuse[actor_id] += 2
        return False, "replay_detected"
    q.append(time.time())
    return True, "ok"


def register_security_event(firebase_db, event: str, payload: dict):
    if firebase_db is None:
        return
    firebase_db.collection("security_audit_logs").add({
        "event": str(event),
        "payload": payload,
        "created_at_ms": int(time.time() * 1000),
    })


def evaluate_abuse_risk(actor_id: str) -> dict:
    score = int(_abuse.get(actor_id, 0))
    risk = "high" if score >= 20 else ("medium" if score >= 10 else "low")
    return {"actor_id": actor_id, "score": score, "risk": risk}


def rotate_security_state(actor_id: str):
    _abuse[actor_id] = max(0, _abuse.get(actor_id, 0) // 2)

File: lib/securityEngine.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { trackEvent } from '@/lib/analytics';

export type SecureAction = {
  action: string;
  actorId: string;
  sessionId: string;
  deviceId: string;
  idempotencyKey?: string;
  timestampMs?: number;
  metadata?: Record<string, unknown>;
};

export type SecurityRiskState = {
  abuseScore: number;
  cooldownUntilMs: number;
  forcedLogout: boolean;
  suspiciousCount: number;
};

const STATE_KEY = 'security_engine_state_v1';
const NONCE_KEY = 'security_engine_nonce_v1';
const ACTION_WINDOW_MS = 60_000;

let state: SecurityRiskState = {
  abuseScore: 0,
  cooldownUntilMs: 0,
  forcedLogout: false,
  suspiciousCount: 0,
};

const recentActions = new Map<string, number>();

function now() { return Date.now(); }

function pruneRecent() {
  const t = now();
  for (const [k, v] of recentActions.entries()) if (t - v > ACTION_WINDOW_MS) recentActions.delete(k);
}

async function persistState() {
  await AsyncStorage.setItem(STATE_KEY, JSON.stringify(state));
}

export async function initSecurityEngine() {
  const raw = await AsyncStorage.getItem(STATE_KEY);
  if (raw) {
    try { state = { ...state, ...(JSON.parse(raw) as Partial<SecurityRiskState>) }; } catch { /* noop */ }
  }
}

export async function validateSecureAction(input: SecureAction): Promise<{ ok: boolean; reason?: string }> {
  const ts = Number(input.timestampMs || now());
  if (Math.abs(now() - ts) > 5 * 60 * 1000) return { ok: false, reason: 'stale_action' };
  if (!input.actorId || !input.sessionId || !input.deviceId) return { ok: false, reason: 'missing_security_context' };
  if (state.forcedLogout) return { ok: false, reason: 'session_invalidated' };
  if (state.cooldownUntilMs > now()) return { ok: false, reason: 'cooldown_active' };

  pruneRecent();
  const replayKey = input.idempotencyKey || `${input.action}:${input.actorId}:${Math.floor(ts / 1000)}`;
  if (recentActions.has(replayKey)) {
    state.abuseScore += 2;
    state.suspiciousCount += 1;
    await persistState();
    return { ok: false, reason: 'duplicate_action' };
  }
  recentActions.set(replayKey, now());
  return { ok: true };
}

export async function registerSecurityEvent(name: string, payload: Record<string, unknown> = {}) {
  const suspicious = name.includes('abuse') || name.includes('replay') || name.includes('denied');
  if (suspicious) {
    state.abuseScore += 1;
    state.suspiciousCount += 1;
    if (state.abuseScore >= 12) state.cooldownUntilMs = now() + 2 * 60 * 1000;
    if (state.abuseScore >= 25) state.forcedLogout = true;
    await persistState();
  }
  trackEvent('custom', { metric: 'security_event', name, suspicious, ...payload });
}

export async function evaluateAbuseRisk(action: string, retryCount = 0) {
  const score = state.abuseScore + retryCount + (action.includes('upload') ? 1 : 0);
  const risk = score >= 20 ? 'high' : score >= 10 ? 'medium' : 'low';
  if (risk !== 'low') await registerSecurityEvent('abuse_risk_eval', { action, risk, score });
  return { risk, score };
}

export function getSecurityRiskState() {
  return { ...state };
}

export async function invalidateSession(reason: string) {
  state.forcedLogout = true;
  await persistState();
  await registerSecurityEvent('session_invalidated', { reason });
}

export async function rotateSecurityState() {
  state.abuseScore = Math.max(0, Math.floor(state.abuseScore / 2));
  state.suspiciousCount = Math.max(0, Math.floor(state.suspiciousCount / 2));
  state.cooldownUntilMs = 0;
  await persistState();
}

export function createAuditEvent(action: string, actorId: string, metadata: Record<string, unknown> = {}) {
  return {
    action,
    actor_id: actorId,
    metadata,
    created_at_ms: now(),
  };
}
